Change export IR to non-functional pre-dispatch IR

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139511

Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some auditing to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

tugsbayasgalan committed:
TEMP enable

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP enable"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP enable"




[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "TEMP enable"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -1200,7 +1200,7 @@ def forward(self, x, y):
 
         model = Foo()
-        inputs = (torch.randn(64), torch.randn(64))
+        inputs = (torch.zeros(64), torch.ones(64))
```
why?

tugsbayasgalan replied:
Likely not needed, this was to make the test deterministic. In very rare cases, they could result in same answer causing the runtime assertion to not fired from max() call.

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -6894,17 +6884,6 @@ def test(ep):
             # running decompositions again should work for all IRs
             ep = export(M(), inp, preserve_module_call_signature=("n",))
             test(ep.run_decompositions({}))
-            if is_training_ir_test(self._testMethodName):
```
Do these tests run in OSS mode? Are we expecting divergence when testing through fbcode?

tugsbayasgalan replied:
In fbcode, we also run the tests with training IR for our own tests: https://www.internalfb.com/diff/D65362160

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -6894,17 +6884,6 @@ def test(ep):
             # running decompositions again should work for all IRs
             ep = export(M(), inp, preserve_module_call_signature=("n",))
             test(ep.run_decompositions({}))
```
previously this would test re-running decompositions, now it won't, right?

tugsbayasgalan replied:
Ahh you are right we are still decomp-ing in training IR case, i will include the code i deleted here.

avikchaudhuri reviewed (commented):
I'd really like a summary here, pointing out the effects on testing. How many variants run on test after this PR? Are there redundant variants? Are we losing coverage on internal workflows that don't switch?

ydwu4 commented on torch/export/exported_program.py:
```diff
@@ -359,11 +359,11 @@ def _is_joint_ir_decomp(ep, joint_loss_index):
                     else:
                         real_script_obj = node.meta["val"].fake_val.real_obj
-                    fake_args.append(maybe_to_fake_obj(fake_mode, real_script_obj))
+                    retracing_args.append(real_script_obj)
```
Why we no longer fakify script objects anymore in re-tracing?

tugsbayasgalan replied:
Because fakify_objects takes in non-fake TOrchbind objects and then fakify them.

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some re-work to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65362160).

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some re-work to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some re-work to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some re-work to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some auditing to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some auditing to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some auditing to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some auditing to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some auditing to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some auditing to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Change export IR to non-functional pre-dispatch IR"


Differential Revision: [D65362160](https://our.internmc.facebook.com/intern/diff/D65362160)

State after this IR:
1. For the tests that require inference IR, they are replaced with ep.run_decomp({}) so export_for_training_run_decomp is sort of redundant but i guess it is still nice that multiple round of retracing still working. In general, we need some auditing to reduce our redundant testing coverages. 
2. After this PR landed and not get reverted for a week or so, i will replace the export_for_training calls with export as they are the same thing now. 
3. Added more tests to also cover now "deprecated" old IR by patching export to use old export. For reviewers, please look at the internal version. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
